src/w3d/renderer/render2dsentence.cpp

```diff
@@ -370,3 +374,559 @@ void FontCharsClass::Free_Character_Arrays()
+
+    while (*text != u'\0') {
+        unichar_t ch = *text++;
+
+        if (ch != u'\n') {
+            extent.X += Font->Get_Char_Spacing(ch);
+        }
+    }
+
+    return extent;
+}
+
+Vector2 Render2DSentenceClass::Get_Formatted_Text_Extents(const unichar_t *text)
+{
+    return Build_Sentence_Not_Centered(text, nullptr, nullptr, false);
+}
+
+void Render2DSentenceClass::Build_Sentence(const unichar_t *text, int *unk1, int *unk2)
+{
+    if (text == nullptr) {
+        return;
+    }
+
+    if (Centered && (WrapWidth > 0.0f || u_strchr(text, u'\n'))) {
+        Build_Sentence_Centered(text, unk1, unk2);
+    } else {
+        Build_Sentence_Not_Centered(text, unk1, unk2);
+    }
+}
+
+void Render2DSentenceClass::Draw_Sentence(uint32_t color)
+{
+    Render2DClass *curr_renderer = nullptr;
+    SurfaceClass *curr_surface = nullptr;
+
+    DrawExtents.Set(0, 0, 0, 0);
+    for (int index = 0; index < SentenceData.Count(); index++) {
+        SentenceDataStruct &data = SentenceData[index];
+
+        if (data.Surface != curr_surface) {
+            curr_surface = data.Surface;
+            bool found = false;
+            for (int renderer_index = 0; renderer_index < Renderers.Count(); renderer_index++) {
+                if (Renderers[renderer_index].Surface == curr_surface) {
+                    found = true;
+                    curr_renderer = Renderers[renderer_index].Renderer;
+                    break;
+                }
+            }
+
+            if (found == false) {
+                curr_renderer = new Render2DClass;
+                curr_renderer->Set_Coordinate_Range(Render2DClass::Get_Screen_Resolution());
+                *curr_renderer->Get_Shader() = Shader;
+                RendererDataStruct render_info;
+                render_info.Renderer = curr_renderer;
+                render_info.Surface = curr_surface;
+                Renderers.Add(render_info);
+                for (int surface_index = 0; surface_index < PendingSurfaces.Count(); surface_index++) {
+                    PendingSurfaceStruct &surface_info = PendingSurfaces[surface_index];
+                    if (surface_info.Surface == curr_surface) {
+                        surface_info.Renderers.Add(curr_renderer);
+                    }
+                }
+            }
+        }
+
+        if (curr_surface == nullptr) {
+            return;
+        }
+
+        SurfaceClass::SurfaceDescription desc;
+        curr_surface->Get_Description(desc);
+        RectClass screen_rect = data.ScreenRect;
+        screen_rect += Location;
+        RectClass uv_rect = data.UVRect;
+        bool add_quad = true;
+
+        if (IsClippedEnabled) {
+            if (screen_rect.right <= ClipRect.left || screen_rect.bottom <= ClipRect.top) {
+                add_quad = false;
+            } else {
+                RectClass clipped_rect;
+                clipped_rect.left = std::max(screen_rect.left, ClipRect.left);
```

Comment:
This exact function is used in a number of places (w3ddisplay draw image for example). We should make a function for it.

Reply:
A function for RectClass perhaps? Any idea what it might be?

Reply:
It's to simplify https://github.com/TheAssemblyArmada/Thyme/blob/200cf1a6178ff0d72a1ef54203f9ac5b22e41a33/src/platform/w3dengine/client/w3ddisplay.cpp#L780
I think you would call this working out an intersection. In other projects I've had an intersection function for my rect classes.

Reply:
Yeah, looks like its intersection, I've finished the sentence class pretty much now, I'll add the intersect function to rectclass as another commit to simplify this bit of code then I think its ready for review and merge.